### PR TITLE
Adds cancel event option for midround random events

### LINKED ISSUE #67962: Dynamic unfavorable events don't have a cancel button for random event spawning
Midrounds spawned by unfavorable events come with a cancel button. Random events don't.

### PR DESCRIPTION
**Pull request by Salex08**
## About The Pull Request

closes: https://github.com/tgstation/tgstation/issues/67962
Adds a Cancel button to stop midround random events

## Why It's Good For The Game
Admins can stop unwanted midround random events

## Changelog

:cl:@Salex08
admin: Admins can now cancel midround random events
/:cl:

### CODE REVIEW

**Review by Mothblocks — CHANGES_REQUESTED** [review]
You need to figure out where `setup()` and stuff are actually called. Signals should only ever be fired in one place.

**Comment by Salex08** [comment]
> You need to figure out where `setup()` and stuff are actually called. Signals should only ever be fired in one place.
does this fix the issue?

File: code/modules/events/_event.dm [hunk 1]
```diff
@@ -93,19 +93,38 @@
...
 	E.current_players = get_active_player_count(alive_check = 1, afk_check = 1, human_check = 1)
 	E.control = src
-	SSblackbox.record_feedback("tally", "event_ran", 1, "[E]")
 	occurrences++
 
 	testing("[time2text(world.time, "hh:mm:ss")] [E.type]")
+	triggering = TRUE
+
+	if (alert_observers)
+		message_admins("Random Event triggering in [RANDOM_EVENT_ADMIN_INTERVENTION_TIME] seconds: [name] (<a href='?src=[REF(src)];cancel=1'>CANCEL</a>)")
```
Comment (Mothblocks) [inline code/modules/events/_event.dm:107]: I don't see where `cancel` is used here.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/modules/events/_event.dm (hunk 1 above)
Comment (Salex08) [reply code/modules/events/_event.dm:107]: its used in topic ref i had tested it out in a local server and it worked to make sure im not mixing it up with prerunevent I used a different message in message_admins

File: code/modules/events/_event.dm [hunk 2]
```diff
@@ -179,6 +198,10 @@
 	if(!processing)
 		return
 
+	if(SEND_GLOBAL_SIGNAL(COMSIG_GLOB_RANDOM_EVENT, src) & CANCEL_RANDOM_EVENT)
+		processing = FALSE
+		kill()
```
[the remark this answers is not in the document]
Comment (Salex08) [reply code/modules/events/_event.dm:203]: true

File: code/modules/events/_event.dm (hunk 1 above)
Comment (Mothblocks) [reply code/modules/events/_event.dm:107]: Ah right

File: code/modules/events/_event.dm [hunk 3]
```diff
@@ -93,19 +93,38 @@
...
 	E.control = src
-	SSblackbox.record_feedback("tally", "event_ran", 1, "[E]")
 	occurrences++
 
 	testing("[time2text(world.time, "hh:mm:ss")] [E.type]")
+	triggering = TRUE
+
+	if (alert_observers)
+		message_admins("Random Event triggering in [RANDOM_EVENT_ADMIN_INTERVENTION_TIME] seconds: [name] (<a href='?src=[REF(src)];cancel=1'>CANCEL</a>)")
+		sleep(RANDOM_EVENT_ADMIN_INTERVENTION_TIME SECONDS)
```
[the remark this answers is not in the document]
Comment (Salex08) [reply code/modules/events/_event.dm]: alright done

File: code/modules/events/_event.dm [hunk 4]
```diff
@@ -179,6 +198,10 @@
 	if(!processing)
 		return
 
+	if(SEND_GLOBAL_SIGNAL(COMSIG_GLOB_RANDOM_EVENT, src) & CANCEL_RANDOM_EVENT)
```
[the remark this answers is not in the document]
Comment (Salex08) [reply code/modules/events/_event.dm:207]: true should have just made a var

File: code/modules/events/_event.dm [hunk 5]
```diff
@@ -93,7 +93,12 @@
 		SSblackbox.record_feedback("tally", "event_admin_cancelled", 1, typepath)
 
 /datum/round_event_control/proc/runEvent(random = FALSE)
-	//We clear our signals first
+/*
+* We clear our signals first so we dont cancel a wanted event by accident,
+* the majority of time the admin will probably want to cancel a single midround spawned random events
+* and not multiple events called by others admins
+* * In the worst case scenario we can still recall a event which we cancelled by accident, which is much better then to have a unwanted event
+*/
```
Comment (Mothblocks) [inline code/modules/events/_event.dm:101]: Move this comment up an indentation level

File: code/modules/events/_event.dm [hunk 6]
```diff
@@ -104,8 +109,8 @@
 	triggering = TRUE
 
 	if (alert_observers)
-		message_admins("Random Event triggering in [RANDOM_EVENT_ADMIN_INTERVENTION_TIME] seconds: [name] (<a href='?src=[REF(src)];cancel=1'>CANCEL</a>)")
-		sleep(RANDOM_EVENT_ADMIN_INTERVENTION_TIME SECONDS)
+		message_admins("Random Event triggering in 10 seconds: [name] (<a href='?src=[REF(src)];cancel=1'>CANCEL</a>)")
+		sleep(100)
```
Comment (Mothblocks) [inline code/modules/events/_event.dm:113]: No no, keep `RANDOM_EVENT_ADMIN_INTERVENTION_TIME` but have that be in seconds.

File: code/modules/events/_event.dm (hunk 5 above)
Comment (Salex08) [reply code/modules/events/_event.dm]: done

File: code/modules/events/_event.dm (hunk 6 above)
Comment (Salex08) [reply code/modules/events/_event.dm]: done like you said

File: code/modules/events/_event.dm [hunk 7]
```diff
@@ -122,6 +147,7 @@
 	var/activeFor = 0 //How long the event has existed. You don't need to change this.
 	var/current_players = 0 //Amount of of alive, non-AFK human players on server at the time of event start
 	var/fakeable = TRUE //Can be faked by fake news event.
+	var/cancel_event = FALSE //Wheter a admin wants this event to be cancelled
```
Comment (Mothblocks) [inline code/modules/events/_event.dm:150]:
```suggestion
	/// Whether a admin wants this event to be cancelled
	var/cancel_event = FALSE
```

File: code/modules/events/_event.dm [hunk 8]
```diff
@@ -93,19 +93,44 @@
...
-	if (alert_observers)
+
+	if(alert_observers)
 		deadchat_broadcast(" has just been[random ? " randomly" : ""] triggered!", "<b>[name]</b>", message_type=DEADCHAT_ANNOUNCEMENT) //STOP ASSUMING IT'S BADMINS!
+
+	SSblackbox.record_feedback("tally", "event_ran", 1, "[E]")
 	return E
 
+//Returns the component for the listener
+/datum/round_event_control/proc/stop_random_event()
```
Comment (Mothblocks) [inline code/modules/events/_event.dm:131]:
```suggestion
/datum/round_event_control/proc/stop_random_event()
	SIGNAL_HANDLER
```

**Comment by Mothblocks** [comment]
There is probably a better way to do this but I think this is fine

**Pull request merged**